File: product-service/src/main/java/product_service/entity/Product.java
package product_service.entity;

import io.swagger.v3.oas.annotations.media.Schema;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import product_service.enums.ProductCategory;

import java.time.LocalDateTime;

@Entity
@Data
@AllArgsConstructor
@NoArgsConstructor
@Schema(description = "Product entity representing food items in the Healthy Food Bank catalog")
public class Product {

    @PrePersist
    protected void onCreate() {
        productAdditionDate = LocalDateTime.now();
        productUpdatedDate = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        productUpdatedDate = LocalDateTime.now();
    }

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Schema(description = "Unique identifier for the product", example = "1", accessMode = Schema.AccessMode.READ_ONLY)
    private int productId;
    
    @Schema(description = "Name of the product", example = "Organic Apples", required = true)
    private String productName;
    
    @Schema(description = "Price of the product", example = "2.99", required = true)
    private double productPrice;
    
    @Schema(description = "Available quantity of the product", example = "100.0", required = true)
    private double productQuantity;
    
    @Schema(description = "Unit of measurement for the product", example = "kg", required = true)
    private String productUnit;

    @Schema(description = "Quantity per unit (e.g., 1 for '1 kg', 250 for '250 ml')", example = "1.0", required = true)
    private double unitQuantity;

    @Schema(description = "Timestamp when the product was added", accessMode = Schema.AccessMode.READ_ONLY)
    private LocalDateTime productAdditionDate;
    
    @Schema(description = "Timestamp when the product was last updated", accessMode = Schema.AccessMode.READ_ONLY)
    private LocalDateTime productUpdatedDate;

    @Column(name = "product_added_by")
    @Schema(description = "ID of the user who added this product", example = "123")
    private Long productAddedBy;

    @Column(name = "vendor_id")
    @Schema(description = "Vendor ID who owns this product", example = "VENDOR001")
    private String vendorId;

    @Column(name = "vendor_name")
    @Schema(description = "Vendor name who owns this product", example = "John Doe")
    private String vendorName;

    @Enumerated(EnumType.STRING)
    @Column(name = "category", length = 20)
    @Schema(description = "Product category", example = "VEGETABLES")
    private ProductCategory category = ProductCategory.OTHERS;

    @Column(name = "delivery_schedule", length = 20)
    @Schema(description = "Delivery schedule for the product", example = "SATURDAY")
    private String deliverySchedule;
}

File: healthy_food_bank_flutter/lib/models/product.dart
enum ProductCategory {
  VEGETABLES,
  FRUITS,
  DAIRY,
  GRAINS,
  PROTEINS,
  BEVERAGES,
  ORGANIC,
  OTHERS,
}

class Product {
  final int id;
  final String name;
  final String? description;
  final double price;
  final String? category;
  final String? vendorId;
  final String? vendorName;
  final int stockQuantity;
  final String? imageUrl;
  final double? unitQuantity;
  final String? productUnit;
  final String? deliverySchedule;
  final bool active;

  Product({
    required this.id,
    required this.name,
    this.description,
    required this.price,
    this.category,
    this.vendorId,
    this.vendorName,
    required this.stockQuantity,
    this.imageUrl,
    this.unitQuantity,
    this.productUnit,
    this.deliverySchedule,
    this.active = true,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: json['productId'] ?? json['id'] ?? 0,
      name: json['productName'] ?? json['name'] ?? '',
      description: json['description'] ?? json['productDescription'],
      price: (json['productPrice'] ?? json['price'] ?? 0).toDouble(),
      category: json['category'] ?? json['productCategory'],
      vendorId: json['vendorId']?.toString() ?? json['productAddedBy']?.toString(),
      vendorName: json['vendorName'],
      stockQuantity: (json['productQuantity'] ?? json['stockQuantity'] ?? 0).toInt(),
      imageUrl: json['imageUrl'],
      unitQuantity: (json['unitQuantity'] ?? 1).toDouble(),
      productUnit: json['productUnit'] ?? 'unit',
      deliverySchedule: json['deliverySchedule'],
      active: json['active'] ?? true,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'productName': name,
      'productPrice': price,
      'productQuantity': stockQuantity,
      'productUnit': productUnit ?? 'unit',
      'unitQuantity': unitQuantity ?? 1,
      if (description != null) 'description': description,
      if (category != null) 'category': category,
      if (imageUrl != null) 'imageUrl': imageUrl,
      if (deliverySchedule != null) 'deliverySchedule': deliverySchedule,
    };
  }

  bool get isInStock => stockQuantity > 10;
  bool get isLowStock => stockQuantity > 0 && stockQuantity <= 10;
  bool get isOutOfStock => stockQuantity <= 0;

  String get priceDisplay => '₹${price.toStringAsFixed(price.truncateToDouble() == price ? 0 : 2)}';

  String get unitDisplay {
    final qty = unitQuantity ?? 0;
    final unit = productUnit ?? 'unit';
    if (qty <= 0 || qty == 1) return unit;
    return '${qty.toStringAsFixed(qty.truncateToDouble() == qty ? 0 : 1)} $unit';
  }

  String get pricePerUnit => '$priceDisplay / $unitDisplay';

  String get stockStatus {
    if (isOutOfStock) return 'Out of Stock';
    if (isLowStock) return 'Low Stock';
    return 'In Stock';
  }
}
